raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+
+            # Set to 0 around the previous point
+            if index != 1:
+                if y_spline[index] < y_spline[index - 1] and y_spline[index-2] < y_spline[index - 1]:
+                    dfdx = 0.
+                    gradient_set_0 = True
+                elif y_spline[index] > y_spline[index - 1] and y_spline[index-2] > y_spline[index - 1]:
+                    dfdx = 0.
+                    gradient_set_0 = True
+            if not gradient_set_0:
+                # Finding the normalised distance x_eff
+                x_eff = (x_spline[index + 1] - x_spline[index - 1])/(x_spline[index + 1] - x_spline[index])
+                if x_eff != 0:
+                    dfdx = (y_spline[index + 1] - y_spline[index - 1])/x_eff
+                else:
+                    raise ZeroDivisionError('Two adjacent spline points have the same x value!')
+        return dfdx
+
+
+cdef class _Extrapolator1D:
+    """
+    Base class for Function1D extrapolators.
+
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param double extrapolation_range: Range covered by the extrapolator. Padded symmetrically to both ends of the input.
+    """
+    typename = NotImplemented
+
+    def __init__(self, double[::1] x, double[::1] f, double extrapolation_range):
+        self._range = extrapolation_range
+        self._x = x
+        self._f = f
+
+    cdef double extrapolate(self, double px, int order, int index, double rx) except? -1e999:
+        raise NotImplementedError(f'{self.__class__} not implemented.')
+
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        raise NotImplementedError(f'{self.__class__} not implemented.')
+
+
+cdef class _ExtrapolatorNone(_Extrapolator1D):
+    """
+    Extrapolator that does nothing.
+    """
+    cdef double extrapolate(self, double px, int order, int index, double rx) except? -1e999:
+        raise ValueError(f'Extrapolation not available. Interpolate within function range {np.min(self._x)}-{np.max(self._x)}.')
+
+    cdef double evaluate(self, double px, int index)  except? -1e999:
+        raise ValueError('Extrapolation not available.')
+
+
+cdef class _Extrapolator1DNearest(_Extrapolator1D):
+    """
+    Extrapolator that returns nearest input value
+    :param object x: 1D array-like object of real values.
+    :param object f: 1D array-like object of real values.
+    :param double extrapolation_range: Range covered by the extrapolator.
+    """
+
+    typename = 'nearest'
+
+    def __init__(self, double [::1] x, double[::1] f, double extrapolation_range):
+        super().__init__(x, f, extrapolation_range)
+        self._last_index = self._x.shape[0] -1
+
+    cdef double evaluate(self, double px, int index) except? -1e999:
+        if px < self._x[0]:
+            return self._f[0]
+        elif px >= self._x[self._last_index]:
+            return self._f[self._last_index]
```

Comment:
Would be good form to add an exception if neither cases fire to indicate the method was given bad data,